File: runtime/engine/go/handler/response.go
package handler

import (
	pb "github.com/socketfunc/faas/runtime/proto"
)

type Response struct {
	conn pb.Internal_PipeServer
}

func (r *Response) Send(topic, event string, id int, payload []byte) error {
	pipeRes := &pb.PipeResponse{
		Version: 1,
		Cmd:     pb.Cmd_Send,
		Packet: &pb.Packet{
			Topic:   topic,
			Event:   event,
			Id:      int32(id),
			Payload: payload,
		},
	}
	if err := r.conn.Send(pipeRes); err != nil {
		return err
	}
	return nil
}

func NewResponse(conn pb.Internal_PipeServer) *Response {
	return &Response{
		conn: conn,
	}
}

File: runtime/engine/nodejs/lib/response.js
class Response {
  constructor (conn) {
    this._conn = conn
  }

  static get CMD () {
    return {
      SEND: 0,
      REPLY: 1,
      BROADCAST: 2
    }
  }

  _send (cmd, topic, event, id, payload) {
    if (!this._conn.writable) {
      return
    }
    const version = 1
    this._conn.write({version, cmd, packet: {topic, event, id, payload}})
  }

  /**
   *
   * @param topic
   * @param event
   * @param id
   * @param payload
   */
  send ({topic, event, id, payload}) {
    this._send(Response.CMD.SEND, topic, event, id, payload)
  }

  close () {
    if (!this._conn.writable) {
      return
    }
    this._conn.end()
  }
}

module.exports = Response
